tests_e2e/tests/agent_update/self_update.py

```diff
@@ -0,0 +1,123 @@
+            else:
+                if agent_source_path.exists():
+                    os.rmdir(agent_source_path)  # Remove if partial build exists
+                source_directory: Path = Path(azurelinuxagent.__path__[0]).parent
+                copy_cmd: str = f"scp -r {source_directory} {agent_source_path}"
+                log.info("Copying agent source %s to %s", source_directory, agent_source_path)
+                run_command(copy_cmd, shell=True)
+                if not agent_source_path.exists():
+                    raise Exception(
+                        f"The agent source was not copied to the expected path {agent_source_path}")
+                version_file: Path = agent_source_path / "azurelinuxagent" / "common" / "version.py"
+                version_cmd = rf"""sed -E -i "s/^AGENT_VERSION\s+=\s+'[0-9.]+'/AGENT_VERSION = '{self._test_version}'/g" {version_file}"""
+                log.info("Setting agent version to %s to build new pkg", self._test_version)
+                run_command(version_cmd, shell=True)
+                makepkg_file: Path = agent_source_path / "makepkg.py"
+                build_cmd: str = f"env PYTHONPATH={agent_source_path} python3 {makepkg_file} -o {agent_source_path}"
+                log.info("Building custom test agent pkg version %s", self._test_version)
+                run_command(build_cmd, shell=True)
+                if not source_pkg_path.exists():
+                    raise Exception(
+                        f"The test pkg was not created at the expected path {source_pkg_path}")
+            target_path: Path = Path("~") / "tmp"
+            log.info("Copying %s to %s:%s", source_pkg_path, self._context.vm, target_path)
+            self._ssh_client.copy_to_node(source_pkg_path, target_path)
+
+    def _verify_agent_updated_to_latest_version(self) -> None:
+        """
+        Verifies the agent updated to latest version from custom test version.
+        We retrieve latest version from goal state and compare with current agent version running as that latest version
+        """
+        latest_version: str = self._ssh_client.run_command("agent_update-self_update_latest_version.py", use_sudo=True)
+        self._verify_guest_agent_update(latest_version)
+
+    def _verify_guest_agent_update(self, latest_version: str) -> None:
+        """
+        Verify current agent version running on latest version
+        """
+
+        def _check_agent_version(latest_version: str) -> bool:
+            waagent_version: str = self._ssh_client.run_command("waagent-version", use_sudo=True)
+            expected_version = f"Goal state agent: {latest_version}"
+            if expected_version in waagent_version:
+                return True
+            else:
+                return False
+
+        waagent_version: str = ""
+        log.info("Verifying agent updated to latest version: {0}".format(latest_version))
```

Comment:
we should also verify that the update was done by 2.8.9.9

Reply:
addressed and checking at L#100